File: solutions/cpp/557/solution.cpp
class Solution {
public:
    string reverseWords(string s) {
        int left = 0;
        for (int right = 0; right <= s.size(); right++) {
            if (right == s.size()) {
                reverse(left, right - 1, s);
            } else if (s[right] == ' ') {
                reverse(left, right - 1, s);
                left = right + 1;
            }
        }

        return s;
    }

    void reverse(int left, int right, string &s) {
        while (left < right) {
            swap(s[left], s[right]);
            left++;
            right--;
        }
    }
};

File: solutions/golang/557/solution.go
func reverseWords(s string) string {
	r := []rune(s)
	left := 0
	for right := 0; right < len(r); right++ {
		if s[right] == ' ' || right == len(r)-1 {
			// Swap characters within the word
			for i, j := left, right-1; i < j; i, j = i+1, j-1 {
				r[i], r[j] = r[j], r[i]
			}
			left = right + 1
		}
	}
	return string(r)
}